src/NodaTime/Xml/XmlSchemaDefinition.cs

```diff
@@ -0,0 +1,238 @@
+                [LocalDateTimeSchemaType] = new[] { localDateTimeRestriction, calendarRestriction },
+                [ZonedDateTimeSchemaType] = new[] { OffsetDateTimeSchemaType, calendarRestriction, zoneRestriction },
+            };
+
+            NodaTimeXmlSchema = CreateNodaTimeXmlSchema();
+        }
+
+        private static readonly XmlSchemaType AnnualDateSchemaType;
+        private static readonly XmlSchemaType DurationSchemaType;
+        private static readonly XmlSchemaType InstantSchemaType;
+        private static readonly XmlSchemaType IntervalSchemaType;
+        private static readonly XmlSchemaType LocalDateSchemaType;
+        private static readonly XmlSchemaType LocalDateTimeSchemaType;
+        private static readonly XmlSchemaType LocalTimeSchemaType;
+        private static readonly XmlSchemaType OffsetSchemaType;
+        private static readonly XmlSchemaType OffsetDateSchemaType;
+        private static readonly XmlSchemaType OffsetDateTimeSchemaType;
+        private static readonly XmlSchemaType OffsetTimeSchemaType;
+        private static readonly XmlSchemaType PeriodBuilderSchemaType;
+        private static readonly XmlSchemaType YearMonthSchemaType;
+        private static readonly XmlSchemaType ZonedDateTimeSchemaType;
+        private static readonly Dictionary<XmlSchemaType, IEnumerable<XmlSchemaType>> DependentSchemaTypes;
+
+        private static XmlQualifiedName AddSchemaType(this XmlSchemaSet schemaSet, XmlSchemaType schemaType)
+        {
+            var schema = new XmlSchema
+            {
+                TargetNamespace = NodaTimeXmlNamespace.Namespace,
+                Items = { schemaType }
+            };
+            if (DependentSchemaTypes.TryGetValue(schemaType, out var dependentSchemaTypes))
+            {
+                foreach (var dependentSchemaType in dependentSchemaTypes)
+                {
+                    schema.Items.Add(dependentSchemaType);
+                }
+            }
+            schemaSet.Add(schema);
+            return schemaType.QualifiedName;
+        }
+
+        private static XmlSchema CreateNodaTimeXmlSchema()
+        {
+            var schemaSetForCollecting = new XmlSchemaSet();
+            var xmlSerializableTypes = typeof(XmlSchemaDefinition).Assembly.GetExportedTypes().Where(t => t.GetInterfaces().Contains(typeof(IXmlSerializable)));
+            foreach (var xmlSerializableType in xmlSerializableTypes)
+            {
+                var schemaProvider = xmlSerializableType.GetCustomAttribute<XmlSchemaProviderAttribute>();
+                if (schemaProvider == null)
+                    throw new InvalidOperationException($"The type '{xmlSerializableType}' is missing the '{nameof(XmlSchemaProviderAttribute)}' on its class definition.");
+                var schemaProviderMethod = xmlSerializableType.GetMethod(schemaProvider.MethodName, BindingFlags.Public | BindingFlags.Static);
```

Comment:
Again, do we actually need to use reflection here? It feels like we could manually call all the methods within this class. I understand we still need the attribute for XmlSchemaExporter, but I don't think we need to use it when we're doing the work ourselves. Unless I've missed something, which is entirely possible.

Reply:
You're absolutely right, we don't need reflection at all. I have removed reflection in 6c06ed466ce2770bc44e2b7d4e0199b807fd2c7d.